docs/cpp/source/tensor_creation.rst

```diff
@@ -0,0 +1,279 @@
+Here is an example of creating a ``TensorOptions`` object that represents a
+64-bit float, strided tensor that requires a gradient, and lives on CUDA device
+1:
+
+.. code-block:: cpp
+
+  auto options =
+    torch::TensorOptions()
+      .dtype(torch::kFloat32)
+      .layout(torch::kStrided)
+      .device({torch::kCUDA, 1})
+      .requires_grad(true);
+
+
+Notice how we use the '"builder"-style methods of ``TensorOptions`` to
+construct the object piece by piece. If we pass this object as the last
+argument to a factory function, the newly created tensor will have these
+properties:
+
+.. code-block:: cpp
+
+  torch::Tensor tensor = torch::full(/*value=*/123, {3, 4}, options);
+
+  assert(tensor.dtype() == torch::kFloat32);
+  assert(tensor.layout() == torch::kStrided);
+  assert(tensor.device().type() == torch::kCUDA);
+  // or
+  assert(tensor.device().is_cuda()); // and !is_cpu()
+  assert(tensor.device().index() == 1);
+  assert(tensor.requires_grad())
+
+  assert(tensor.options() == options)
+
+Now, you may be thinking: do I really need to specify each axis for every new
+tensor I create? Fortunately not, as every axis has a default value. These
+defaults are:
+
+- `kFloat32` for the dtype,
+- `kStrided` for the layout,
+- `kCPU` for the device,
+- `false` for `requires_grad`.
+
+What this means is that any axis you omit during the construction of a
+``TensorOptions`` object will take on its default value. For example, this is
+our previous ``TensorOptions`` object, but with the ``dtype`` and ``layout``
+defaulted:
+
+.. code-block:: cpp
+
+  auto options = torch::TensorOptions().device({torch::kCUDA, 1}).requires_grad(true);
+
+In fact, we can even omit all axes to get an entirely defaulted
+``TensorOptions`` object:
+
+.. code-block:: cpp
+
+  auto options = torch::TensorOptions(); // or `torch::TensorOptions options;`
+
+A nice consequence of this is that the ``TensorOptions`` object we just spoke
+so much about can actually be entirely omitted from any tensor factory call:
+
+.. code-block:: cpp
+
+  // A 32-bit float, strided, CPU tensor that does not require a gradient.
+  torch::Tensor tensor = torch::randn({3, 4});
+  torch::Tensor range = torch::arange(5, 10);
+
+This is quite exciting. But wait, there's more. In the API presented here so
+far, you may have noticed that the initial ``torch::TensorOptions()`` is quite
+a mouthful to write. The good news, is that for every construction axis (dtype,
+layout, device and ``requires_grad``), there is one *free function* in the
+``torch::`` namespace which you can pass a value for that axis, and which then
+returns a ``TensorOptions`` object preconfigured with that axis, but allowing
+even further modification via the builder-style methods shown above. For example,
+
+.. code-block:: cpp
+
+  torch::ones(10, torch::TensorOptions().dtype(torch::kFloat32))
+
+is equivalent to
+
+.. code-block:: cpp
+
+  torch::ones(10, torch::dtype(torch::kFloat32))
```

Comment:
Would it be useful to mention that you can actually use the constructor of TensorOptions to even further shorten this to ```torch::ones(10, torch::kFloat32)``` ?

Reply:
Oh yes absolutely, thanks